src/compiler/core.rs

```diff
@@ -1,13 +1,42 @@
-pub fn for_each<TItem, TReturn>(
-    array: &[TItem],
-    callback: &mut dyn FnMut(&TItem, usize) -> Option<TReturn>,
+use crate::SortedArray;
+
+pub fn for_each<
+    TCollection: IntoIterator,
+    TReturn,
+    TCallback: FnMut(TCollection::Item, usize) -> Option<TReturn>,
+>(
+    array: TCollection,
+    mut callback: TCallback,
 ) -> Option<TReturn> {
     array
-        .iter()
+        .into_iter()
         .enumerate()
         .find_map(|(index, item)| callback(item, index))
 }
 
+fn some<TItem>(array: &[TItem], predicate: Option<Box<dyn FnMut(&TItem) -> bool>>) -> bool {
+    predicate.map_or(!array.is_empty(), |predicate| array.iter().any(predicate))
+}
+
+pub fn concatenate<TItem>(mut array1: Vec<TItem>, mut array2: Vec<TItem>) -> Vec<TItem> {
+    if !some(&array2, None) {
+        return array1;
+    }
+    if !some(&array1, None) {
+        return array2;
+    }
+    array1.append(&mut array2);
+    array1
+}
+
+pub fn insert_sorted<TItem>(array: &mut SortedArray<TItem>, insert: TItem) {
```

Comment:
Oh wait looks like I already do have a `SortedArray` type (which means the Typescript version does, so I wonder why the one I commented on above isn't taking a `SortedArray` based on its name?)

Reply:
This seems like it should be a function on `SortedArray`.